plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+PlutusTx.makeLift ''Sqrt
+
+{-# INLINABLE rsqrt #-}
+rsqrt :: Integer -> Integer -> Sqrt
+rsqrt n d
+    | n * d < 0 = Imaginary
+    | n == 0    = Exact 0
+    | n == d    = Exact 1
+    | n < 0     = rsqrt (negate n) (negate d)
+    | otherwise = go 1 $ 1 + divide n d
+  where
+    go :: Integer -> Integer -> Sqrt
+    go l u
+        | l * l * d == n = Exact l
+        | u == (l + 1)   = Irrational l
+        | otherwise      =
+              let
+                m = divide (l + u) 2
+              in
+                if m * m * d <= n then go m u
+                                  else go l m
+
+{-# INLINABLE isqrt #-}
+isqrt :: Integer -> Sqrt
+isqrt n = rsqrt n 1
+
+{-# INLINABLE calculateInitialLiquidity #-}
+calculateInitialLiquidity :: Integer -> Integer -> Integer
+calculateInitialLiquidity outA outB = case isqrt (outA * outB) of
+    Exact l
+        | l > 0 -> l
+    Irrational l
+        | l > 0 -> l + 1
+    _           -> traceError "insufficient liquidity"
+
+{-# INLINABLE calculateAdditionalLiquidity #-}
+calculateAdditionalLiquidity :: Integer -> Integer -> Integer -> Integer -> Integer -> Integer
+calculateAdditionalLiquidity oldA oldB liquidity delA delB = case rsqrt (liquidity * liquidity * newProd) oldProd of
+    Imaginary    -> traceError "insufficient liquidity"
+    Exact x      -> x - liquidity
+    Irrational x -> x - liquidity
+  where
+    oldProd, newProd :: Integer
+    oldProd = oldA * oldB
+    newProd = (oldA + delA) * (oldB + delB)
+
+{-# INLINABLE calculateRemoval #-}
+calculateRemoval :: Integer -> Integer -> Integer -> Integer -> (Integer, Integer)
+calculateRemoval inA inB liquidity diff = (f inA, f inB)
+  where
+    f :: Integer -> Integer
+    f x = x - divide (x * diff) liquidity
+
+data LiquidityPool = LiquidityPool
+    { lpCoinA :: Coin
+    , lpCoinB :: Coin
+    } deriving (Show, Generic, ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''LiquidityPool
+PlutusTx.makeLift ''LiquidityPool
+
+instance Eq LiquidityPool where
+    {-# INLINABLE (==) #-}
+    x == y = (lpCoinA x == lpCoinA y && lpCoinB x == lpCoinB y) ||
+             (lpCoinA x == lpCoinB y && lpCoinB x == lpCoinA y)
+
+{-# INLINABLE hashLiquidityPool #-}
+hashLiquidityPool :: LiquidityPool -> ByteString
+hashLiquidityPool LiquidityPool{..} = sha2_256 $ concatenate (hashCoin c) (hashCoin d)
+  where
+    (c, d)
+        | lpCoinA `coinLT` lpCoinB = (lpCoinA, lpCoinB)
+        | otherwise                = (lpCoinB, lpCoinA)
+
+newtype Uniswap = Uniswap
+    { usCoin :: Coin
+    } deriving stock    (Show, Generic)
+      deriving anyclass (ToJSON, FromJSON, ToSchema)
+
+PlutusTx.makeLift ''Uniswap
+
+instance Prelude.Eq Uniswap where
+    u == v = usCoin u Prelude.== usCoin v
+
+instance Prelude.Ord Uniswap where
+    compare u v = Prelude.compare (usCoin u) (usCoin v)
+
+data UniswapAction = Create LiquidityPool | Close | Swap | Remove | Add
+    deriving Show
+
+PlutusTx.unstableMakeIsData ''UniswapAction
```

Comment:
It's better to use `makeIsDataIndexed` with indices for the constructors (I know this is just an example but it's better to do it right because people are going to copy this)